src/backend/distributed/planner/insert_select_planner.c

```diff
@@ -726,7 +752,8 @@ RouterModifyTaskForShardInterval(Query *originalQuery,
 	 * prevent shard pruning logic (i.e, namely UpdateRelationNames())
 	 * modifies range table entries, which makes hard to add the quals.
 	 */
-	if (!allReferenceTables)
+	RTEListProperties *rteListProperties = GetRTEListPropertiesForQuery(copiedSubquery);
```

Comment:
`GetRTEListPropertiesForQuery ` is not a free to call  function, does lots of traversals. And, the input query is not changing on every iteration. Can we move `GetRTEListPropertiesForQuery ` out of the loop?

Reply:
Actually currently it's already out of the loop in `RouterModifyTaskForShardInterval`.

Do you mean, I should pull this function call to `CreateDistributedInsertSelectPlan` and pass it to `RouterModifyTaskForShardInterval` as another parameter ?

Reply:
ah, sorry I misread, we can ignore this comment